### DISCUSSION TITLE
infinite renders / history.replaceState calls on error in dev (when running as non-standard spa)

### DISCUSSION THREAD

**Discussion by gkemmey** [vercel/next.js#44855] [Help]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

```
yarn run v1.22.10

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 20.6.0: Wed Jan 12 22:22:45 PST 2022; root:xnu-7195.141.19~2/RELEASE_ARM64_T8101
    Binaries:
      Node: 16.13.0
      npm: 8.1.0
      Yarn: 1.22.10
      pnpm: N/A
    Relevant packages:
      next: 12.1.6-canary.6
      react: 17.0.1
      react-dom: 17.0.1

✨  Done in 1.05s.
```

### What browser are you using? (if relevant)

chrome 100.0.4896.88 (Official Build) (arm64) | safari Version 15.4 (16613.1.17.1.13, 16613)

### How are you deploying your application? (if relevant)

not deploying, running `next dev`

### Describe the Bug

disclaimer: i'm trying to run a react-router spa inside the nextjs framework according to the [this gist](https://gist.github.com/tannerlinsley/65ac1f0175d79d19762cf06650707830) and [this article](https://colinhacks.com/essays/building-a-spa-with-nextjs). the goal is to migrate a large create-react-app-style spa one page at a time.

the app appears to infinitely render when there's a client-side runtime error on dev while using rewrites in `next.config.js`. i've captured some screenshots.

| browser | screenshot | notes |
|-|-|-|
| chrome | <img width="1299" alt="Screen Shot 2022-04-23 at 7 35 47 PM" src="https://user-images.githubusercontent.com/1423258/164951270-f260d27b-c28d-49d2-a447-593e747621cd.png"> | kinda hard to see, but in the console badges there's been 12000+ errors. the tab eventually crashes. |
| safari | <img width="1267" alt="Screen Shot 2022-04-23 at 7 35 17 PM" src="https://user-images.githubusercontent.com/1423258/164951331-8e5ba6fb-7429-4455-b092-427b7667153d.png"> | the last error message from safari is more helpful indicating it has blocked `history.replaceState` |

in `next/dist/client/index.js`, i can seemingly prevent the error in `Container#componentDidMount` by only calling `router.replace` if it looks like we're gonna change the path. 

idk if that's actually helpful. i'm mostly fumbling at the keyboard looking for a root cause. but still, here's a screenshot of that "fix" (manually editing node_modules) and a link to where i applied the fix on github:

| screenshot | github link |
| - | - |
| <img width="776" alt="Screen Shot 2022-04-23 at 8 37 46 PM" src="https://user-images.githubusercontent.com/1423258/164951434-4b28f98f-a85d-4fd2-831c-b78dfa648a79.png"> | [link](https://github.com/vercel/next.js/blob/v12.1.6-canary.6/packages/next/client/index.tsx#L122) |

### Expected Behavior

should show the development error page from `ReactDevOverlay`.

i'm certainly aware the spa setup isn't really supported. be nice if it worked for migrating. be nice if the error overlay worked. will totally understand if this gets closed because it's not worth fixing / supporting.

### To Reproduce

sample app: https://github.com/gkemmey/nextjs-react-router. 

i forked the example in the linked article above, updated next, react, and react-router. the app triggers a superficial runtime error by calling `console.foo()` in render of `pages/index.js`, and it has a superficial rewrite (to a path never used). 

as is, it should do the infinite rendering running `yarn dev`. if you ditch the rewrites, it should show the error overlay again.

**Comment by balazsorban44** [comment]
Have you been able to check out:

- https://nextjs.org/docs/migrating/from-create-react-app
- https://nextjs.org/docs/migrating/from-react-router

**Comment by gkemmey** [comment]
hey, thanks for looking!

yes, but unless i missed it they don't really cover incrementally migrating. great steps for migrating wholesale, but i was hoping to avoid that.

**Comment by jankaifer** [comment]
Hi, this sounds useful, but we don't have enough manpower to look into this. Since it's not really officially supported.
I'll convert this into a discussion for now.

Have you considered using `MemoryRouter` or `HashRouter`?
